design_docs/analysis_task_template.md

```diff
@@ -0,0 +1,742 @@
+        if not self.config.has_option(section, 'endDate'):
+            self.config.set(section, 'endDate', endDate)
+
+        # }}}
+
+    def run(self):  # {{{
+        """
+        Runs the analysis task.
+
+        Individual tasks (children classes of this base class) should first
+        call this method to perform any common steps in an analysis task,
+        then, perform the steps required to run the analysis task.
+
+        Authors
+        -------
+        <List of authors>
+
+        Last Modified
+        -------------
+        <MM/DD/YYYY>
+        """
+
+        # here is where the main "meat" of the analysis task goes
+
+        self._my_sub_task('someText', arg2='differentText')
+        return
+        # }}}
+
+    # here is where you add helper methods that are meant to be non-public
+    # (they start with an underscore), meaning you don't expect anyone to
+    # access them outside of this file.  Typically you won't put as much in
+    # the docstring as you would for a public function or method.
+    #
+    # you can either pass arguments (with or without defaults) or you can
+    # "save" arguments as member variables of `self` and then get them back
+    # (like `self.myArg` here).
+    def _my_sub_task(self, arg1, arg2=None):  # {{{
+        """
+        <Performs my favority subtask>
+        """
+
+        # perform the task
+        print 'myArg:', self.myArg
+        print 'arg1:', arg1
+        if arg2 is not None:
+            print 'arg2:', arg2
+        # }}}
+
+# }}}
+
+# vim: foldmethod=marker ai ts=4 sts=4 et sw=4 ft=python
+```
+
+<h4> Implementation: Validation within Analysis Tasks <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+
+Here is an example (from `ocean.climatology_map.ClimatologyMap`) of what 
+the new `__init__` and `setup_and_check` methods :
+```python
+def __init__(self, config, streamMap=None, variableMap=None,
+             fieldName=None):  # {{{
+    """
+    Construct the analysis task.
+
+    Parameters
+    ----------
+    config :  instance of MpasAnalysisConfigParser
+        Contains configuration options
+
+    streamMap : dict, optional
+        A dictionary of MPAS-O stream names that map to their mpas_analysis
+        counterparts.
+
+    variableMap : dict, optional
+        A dictionary of MPAS-O variable names that map to their
+        mpas_analysis counterparts.
+
+    fieldName : {'sst', 'mld', 'sss'}
+        The name of the field to be analyzed
+
+    Raises
+    ------
+    ValueError : if `fieldName` is not provided or is not one of the
+        supported values
+
+    Authors
+    -------
+    Xylar Asay-Davis
+
+    Last Modified
+    -------------
+    03/16/2017
+    """
+    # first, call the constructor from the base class (AnalysisTask)
+    AnalysisTask.__init__(config, streamMap, variableMap)
+
+    upperFieldNames = {'sst': 'SST',
+                       'mld': 'MLD',
+                       'sss': 'SSS'
```

Comment:
that sounds good. I was trying to figure out where it would be easier to start in reviewing these. I will hold on for now, and then start straight from #128 and #129 when you give us the go ahead.